File: python-fp/app/chess/GameCommands.py
from dataclasses import dataclass
from typing import Optional, List

from toolz import compose

from app.chess.Pieces import PieceType
from app.chess.Position import Position, fromString
from app.chess.utils import Maybe, Nothing, Just, Either, Right, Left, mapl, when, bind, maybeToLeft


class GameCommands:
    pass


@dataclass
class Pick(GameCommands):
    position: Position


@dataclass
class Move(GameCommands):
    fromPos: Position
    toPos: Position


@dataclass
class ShowBoard(GameCommands):
    pass


@dataclass
class Promotion(GameCommands):
    position: Position
    pieceType: PieceType


def trim(s: str):
    return s.strip()


def parsePromoteCommand(args) -> Either[List[str], GameCommands]:
    if len(args) != 2:
        return Left(["If you are trying to promote a pawn, type promote <position> <piece name>"])

    pos = fromString(args[0])

    def piece():
        return {
            "rook": Just(PieceType.Rook),
            "queen": Just(PieceType.Queen),
            "bishop": Just(PieceType.Bishop),
            "knight": Just(PieceType.Knight)
        }.get(args[1], Nothing)

    return pos.map(lambda p: piece().map(lambda x: Right(Promotion(p, x)))
                   .orElse(Left(["Invalid piece. You can promote to queen, rook, bishop, or knight"]))
                   ).orElse(Left(["Invalid position. First argument to promote must be a position e.g. a1"]))


def parsePickCommand(pos: str):
    return fromString(pos).map(lambda x: Right(Pick(x))) \
        .orElse(Left(["invalid position. If you want to highlight the moves from a position, "
                      "give just a position e.g. a1"]))


def parseMoveCommand(fromPosStr: str, toPosStr: str):
    msg = ["invalid position. If you are trying to move a piece, give two positions separated by space e.g a1 a2"]
    # return fromString(fromPosStr).map(lambda x: fromString(toPosStr).map(lambda y: Right(Move(x, y)))
    #                                                                 .orElse(Left(msg))
    #                                   ).orElse(Left(msg))
    from_ = maybeToLeft(fromString(fromPosStr), ["Invalid source position!"])
    to_   = maybeToLeft(fromString(toPosStr), ["Invalid target position!"])

    res = from_ |bind| (lambda x:
                        to_ |bind| (lambda y:
                                    Right(Move(x, y))))
    return res


def parseInput(inputStr: Maybe[str]) -> Either[List[str], GameCommands]:
    def handleCmd(args):
        if args[0] == "promote":
            return parsePromoteCommand(args[1:])
        if len(args) == 1:
            return parsePickCommand(args[0])
        if len(args) == 2:
            return parseMoveCommand(args[0], args[1])
        return Left(["Unrecognised command"])

    return inputStr.map(trim).filter(lambda x: x) \
        .map(lambda x: x.split(" ")) \
        .map(mapl(trim)) \
        .map(handleCmd) \
        .orElse(Right(ShowBoard()))

File: typescript-fp/src/chess/GameCommands.ts
import {List} from 'immutable';
import Record from 'dataclass';
import * as M from 'pattern-matching-ts/lib/match'
import { pipe } from 'fp-ts/lib/function'
// import * as O from 'fp-ts/lib/Option'
import { none, some, fromPredicate, Option, map, filter, getOrElse } from 'fp-ts/Option'
import {fromString, Position} from "./Position";
import {charRangeInclusive} from "./utils";
import {Either, left, right} from "fp-ts/Either";
import {BackgroundColor} from "./ChessBoard";


// class GameCommands {
//
// }

// class Pick extends Record<GameCommands> {
//     position: Position;
// }
//
// class Move extends Record<GameCommands> {
//     fromPos: Position;
//     toPos: Position;
// }


// const Pick = Record({ position:   })
// const Move = Record({ a: 1, b: 2 })

// const GameCommands: any = Pick | Move

interface Pick {
    readonly _tag: 'Pick'
    readonly position: Position
}

interface Move {
    readonly _tag: 'Move'
    readonly fromPosition: Position
    readonly toPosition: Position
}

interface ShowBoard {
    readonly _tag: 'ShowBoard'
}

//This just means wherever I am expecting GameCommands I can pass Pick, Move
type GameCommands = Pick | Move | ShowBoard



const pickCmd = (pos: Position): GameCommands => ({_tag: "Pick", position: pos})
const moveCmd = (from: Position, to: Position): GameCommands => ({_tag: 'Move', fromPosition: from, toPosition: to})
const showBoardCmd = (): GameCommands => ({_tag: "ShowBoard"})


//export declare const map: <A, B>(f: (a: A) => B) => (fa: Option<A>) => Option<B>

function trim(str: string) {
    return str.trim()
}

const splitIntoWords = (sep: string) => (str: string) => str.split(sep).map(it => it.trim()).filter(it => it != "")


function parsePickCommand(posStr: string) {
    return pipe(fromString(posStr),
        map(p => right(pickCmd(p))),
        getOrElse(()=> left("invalid position. If you want to highlight the moves from a position, give just a position e.g. a1"))
    )
}

function parseMoveCommand(fromPosStr: string, toPosStr: string) {
    let msg = "invalid position. If you are trying to move a pice give two positions separaated by space e.g a1 a2"
    function seeIfToPositionIsValid(x: Position) {
        return pipe(
            fromString(toPosStr),
            map(y => right(moveCmd(x, y))),
            getOrElse(()=> left(msg))
        )
    }

    return pipe(fromString(fromPosStr),
        map(x => seeIfToPositionIsValid(x)),
        getOrElse(()=> left(msg))
    )
}

export function parseInput(inputStr: Option<string>): Either<string, GameCommands> {
    function handleCommand(args: string[]) {
        if(args[0] == "promote") return left("Unimplemented")
        if(args.length == 1) return parsePickCommand(args[0])
        if(args.length == 2) return parseMoveCommand(args[0], args[1])
        return left("Unrecognised command")
    }

    //     .map(lst => lst.map(trim))
    return pipe(inputStr,
                  map(trim),
                  filter(x => x != ""),
                  map(splitIntoWords(" ")),
                  map(handleCommand),
                  getOrElse(() => right(showBoardCmd()))
    )
}


export {GameCommands, Move, Pick, pickCmd, moveCmd}